server-spi-private/src/main/java/org/keycloak/models/session/PersistentUserSessionAdapter.java

```diff
@@ -145,6 +147,11 @@ public String getIpAddress() {
         return getData().getIpAddress();
     }
 
+    @Override
+    public DeviceInfo getDeviceInfo() {
```

Comment:
For example here. If this is kept inside `notes`, there is no need to change this class.